File: ruby/graphs/depth_first_search.rb
module Graphs
  class Node
    attr_accessor :vertex, :next_pointer

    def initialize(vertex, next_pointer)
      @vertex, @next_pointer = vertex, next_pointer
    end
  end

  class DepthFirstSearch
    class << self
      @time = nil
      @found = nil
      @topological_list = nil

      # Internal: Traverses through the vertices of a graph in a DEPTH wise manner
      #           Adjacent node is considered only after traversing the complete
      #           depth of the current node
      #
      # graph - Graph to be searched
      # searchable - Node in the graph to be searched for
      #
      # Examples
      #   DFS(graph, graph.vertices.last)
      #
      def DFS(g, searchable = nil, topological = false)
        g.vertices.each do |u|
          u.color = 'WHITE'
          u.pi = nil
        end
        @time = 0

        g.vertices.each do |u|
          DFS_visit(g, u, searchable, topological) if u.color == 'WHITE'
        end
        @topological_list || @found
      end

      def DFS_visit(g, u, searchable = nil, topological = false)
        @time += 1
        u.d = @time
        u.color = 'GRAY'
        u.adj_list.each do |v|
          if v.color == 'WHITE'
            v.pi = u
            @found = "FOUND!!" if searchable && v.attribute == searchable.attribute
            DFS_visit(g, v, searchable, topological)
          end
        end
        u.color = 'BLACK'
        @time += 1
        u.f = @time
        if topological
          new_node = Node.new(u, nil)
          new_node.next_pointer = @topological_list
          @topological_list = new_node
        end
      end

    end
  end
end

File: Java/GraphAlgorithms/DepthFirstSearch.java
package graphalgorithms;

import java.util.List;

public class DepthFirstSearch {

    private int time;
    
    public DepthFirstSearch() {
        this.time = 0;
    }

    public void dfs(Graph g) {
        if (g == null) {
            throw new IllegalArgumentException("Graph object null");
        }
        System.out.printf("DFS: ");
        List<Vertex> vertices = g.getVertices();
        for (Vertex u : vertices) {
            u.color = Color.WHITE;
            u.p = null;
            u.d = -1;
            u.f = -1;
        }
        time = 0;
        for (Vertex u : vertices) {
            if (u.color == Color.WHITE) {
                dfsVisit(g, u);
            }
        }
        System.out.printf("%n");
        System.out.printf("Total time: %d%n", time);
    }
    
    protected void dfsVisit(Graph g, Vertex u) {
        if (g == null || u == null) {
            throw new IllegalArgumentException("Graph object or Vertex object null");
        }
        time = time + 1;
        u.d = time;
        u.color = Color.GRAY;
        System.out.printf("%s ", u);
        List<Vertex> adjVertices = u.neighbors;
        for (Vertex v : adjVertices) {
            if (v.color == Color.WHITE) {
                v.p = u;
                dfsVisit(g, v);
            }
        }
        u.color = Color.BLACK;
        time = time + 1;
        u.f = time;
    }
}